archive/miu_language/decision_nec.lean

```diff
@@ -0,0 +1,310 @@
+/-
+Copyright (c) 2020 Gihan Marasingha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gihan Marasingha
+-/
+import .basic
+import data.nat.modeq
+import tactic.ring
+
+/-!
+# Decision procedure - necessary condition
+
+We introduce a condition `decstr` and show that if a string `en` is `derivable`, then `decstr en`
+holds.
+
+Using this, we give a negative answer to the question: is `"MU"` derivable?
+-/
+
+namespace miu
+
+open miu_atom
+
+/-!
+### Numerical condition on the `I count
+
+Let `icount st` denote the number of `I`s in `st : miustr`. We'll show that the `icount` of a
+derivable string must be 1 or 2 moduloe 3. To do this, it suffices to show that if the `miustr`
+`en` is derived from `st`, then `icount en` moudulo 3 is either equal to or is twice `icount st`
+modulo 3.
+-/
+
+/--
+`icount` is the number of `I`s in an `miustr`
+-/
+def icount : miustr → ℕ
+| [] := 0
+| (I::cs) := 1 + icount cs
+| (_::cs) := icount cs
```

Comment:
Isn't there a built-in `count` function? Why not use that, and reuse a mathlib lemma about `count` and `append`?

Reply:
I've made the suggested change. Much improved code follows. Thanks!